### PR TITLE
feat(data_partition): added data partition rules and tests

### PR DESCRIPTION
**Pull request by NSSPKrishna**
# Description
 added data partition rules and tests

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist:

Please delete options that are not relevant.

- [x] My commit message follows [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] My code is formatted to [Go standards](https://go.dev/blog/gofmt)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes. Go [here](https://github.com/newrelic/terraform-provider-newrelic#testing) for instructions on running tests locally.

### CODE REVIEW

File: newrelic/resource_newrelic_data_partition.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+	if created == nil {
+		return diag.Errorf("err: data partition rule create result wasn't returned or rule was not created.")
+	}
+
+	ruleID := created.Rule.ID
+
+	d.SetId(ruleID)
+
+	//Need retry mechanism
+	retryErr := resource.RetryContext(ctx, d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
```
Comment (sanderblue) [inline newrelic/resource_newrelic_data_partition.go:147]: Just curious, do we need the retry due to async latency behavior?

File: newrelic/resource_newrelic_data_partition.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+				Description: "The name of the data partition where logs will be allocated once the rule is enabled.",
+				Required:    true,
+				ForceNew:    true,
+				ValidateDiagFunc: func(v any, p cty.Path) diag.Diagnostics {
+					value := v.(string)
+					var diags diag.Diagnostics
+					if !strings.HasPrefix(value, "Log_") {
+						diag := diag.Diagnostic{
+							Severity: diag.Error,
+							Summary:  "INVALID_DATA_PARTITION_INPUT",
```
[the remark this answers is not in the document]
Comment (NSSPKrishna) [reply newrelic/resource_newrelic_data_partition.go:69]: Yes, it is.

File: newrelic/resource_newrelic_data_partition.go (hunk 1 above)
Comment (NSSPKrishna) [reply newrelic/resource_newrelic_data_partition.go:147]: Yes

File: newrelic/resource_newrelic_data_partition.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sanderblue) [reply newrelic/resource_newrelic_data_partition.go:69]: If the API does the validation for us, we shouldn't have validation in the provider. For example, what if the validation changes on the API side? Now we have to change it in our code too.

The API should be responsible for this type of validation 🙂

[the remark this answers is not in the document]
Comment (NSSPKrishna) [reply newrelic/resource_newrelic_data_partition.go]: Removed the terraform validation part.

**Review by sanderblue — APPROVED** [review]

**Pull request merged**